plugins/inputs/suricata/suricata.go

```diff
@@ -0,0 +1,223 @@
+
+  # Delimiter for flattening field keys, e.g. subitem "alert" of "detect"
+  # becomes "detect_alert" when delimiter is "_".
+  delimiter = "_"
+`
+
+// SampleConfig returns a sample TOML section to illustrate configuration
+// options.
+func (s *Suricata) SampleConfig() string {
+	return sampleConfig
+}
+
+// Start initiates background collection of JSON data from the socket
+// provided to Suricata.
+func (s *Suricata) Start(acc telegraf.Accumulator) error {
+	var err error
+	if s.inputListener == nil {
+		s.inputListener, err = net.ListenUnix("unix", &net.UnixAddr{
+			Name: s.Source,
+			Net:  "unix",
+		})
+		if err != nil {
+			return err
+		}
+		ctx, cancel := context.WithCancel(context.Background())
+		s.cancel = cancel
+		s.inputListener.SetUnlinkOnClose(true)
+		s.wg.Add(1)
+		go func() {
+			defer s.wg.Done()
+			go s.handleServerConnection(ctx, acc)
+		}()
+	}
+	return nil
+}
+
+// Stop causes the plugin to cease collecting JSON data from the socket provided
+// to Suricata.
+func (s *Suricata) Stop() {
+	s.inputListener.Close()
+	if s.cancel != nil {
+		s.cancel()
+	}
+	s.wg.Wait()
+	s.inputListener = nil
+}
+
+func (s *Suricata) readInput(ctx context.Context, acc telegraf.Accumulator, conn net.Conn) {
+	reader := bufio.NewReaderSize(conn, InBufSize)
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		default:
+			line, rerr := reader.ReadBytes('\n')
+			if rerr == io.EOF {
+				return
+			} else if len(line) > 0 {
+				s.parse(acc, line)
+			}
+		}
+	}
+}
+
+func (s *Suricata) handleServerConnection(ctx context.Context, acc telegraf.Accumulator) {
+	var err error
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		default:
+			var conn net.Conn
+			conn, err = s.inputListener.Accept()
+			if err != nil {
+				if !strings.HasSuffix(err.Error(), ": use of closed network connection") {
+					acc.AddError(err)
+				}
+				return
+			}
+			s.readInput(ctx, acc, conn)
+		}
+	}
+}
+
+func flexFlatten(outmap map[string]interface{}, field string, v interface{}, delimiter string) error {
+	switch t := v.(type) {
+	case map[string]interface{}:
+		for k, v := range t {
+			var err error
+			if field == "" {
+				err = flexFlatten(outmap, k, v, delimiter)
+			} else {
+				err = flexFlatten(outmap, fmt.Sprintf("%s%s%s", field, delimiter, k), v, delimiter)
+			}
+			if err != nil {
+				return err
+			}
+		}
+	case float64:
+		outmap[field] = v.(float64)
+	default:
+		return fmt.Errorf("unsupported type %T encountered", t)
+	}
+	return nil
+}
+
+func (s *Suricata) parse(acc telegraf.Accumulator, sjson []byte) {
+	// initial parsing
+	var result map[string]interface{}
+	err := json.Unmarshal([]byte(sjson), &result)
+	if err != nil {
+		acc.AddError(err)
+		return
+	}
+
+	// check for presence of relevant stats
+	if _, ok := result["stats"]; !ok {
+		acc.AddError(fmt.Errorf("input does not contain necessary 'stats' sub-object"))
+		return
+	}
+
+	if _, ok := result["stats"].(map[string]interface{}); !ok {
```

Comment:
This is minor, but what could be done to simplify this is have a result type that is partially specified:
```
type StatsEvent struct {
    Stats map[string]interface{} `json:"stats"`
}
```

Then checking for this error will be handled by Unmarshal.

Reply:
I must say that I did the explicit checks on purpose as they allow for more precise messages than just having the unmarshaling fail. If you consider this to be a potential performance impact (unlikely IMHO) please say so and I will change it, otherwise I would prefer to keep it like this.